backend/messaging/wizard_topic.go

```diff
@@ -1,41 +1,92 @@
 package messaging
 
-import "sync"
+import (
+	"context"
+	"errors"
+	"strconv"
+	"sync"
+
+	"github.com/sensu/sensu-go/types"
+)
 
 // WizardTopic encapsulates state around a WizardBus topic and its
 // consumer channel bindings.
 type WizardTopic struct {
+	ID       string
+	Bindings map[int64]Subscriber
+	Ring     types.Ring
+
+	counter int64
 	sync.RWMutex
-	Bindings map[string]chan<- interface{}
 }
 
 // Send a message to all subscribers to this topic.
 func (wTopic *WizardTopic) Send(msg interface{}) {
 	wTopic.RLock()
 	defer wTopic.RUnlock()
 
-	for _, ch := range wTopic.Bindings {
+	for _, subscriber := range wTopic.Bindings {
 		select {
-		case ch <- msg:
+		case subscriber.Receiver() <- msg:
 		default:
 			continue
 		}
 	}
 }
 
-// Subscribe a channel, identified by a consumer name, to this topic.
-func (wTopic *WizardTopic) Subscribe(consumer string, channel chan<- interface{}) {
-	wTopic.Lock()
-	wTopic.Bindings[consumer] = channel
-	wTopic.Unlock()
+// SendDirect sends a message directly to a subscriber of this topic.
+func (wTopic *WizardTopic) SendDirect(msg interface{}) error {
+	if wTopic.Ring == nil {
+		return errors.New("no ring for topic: " + wTopic.ID)
+	}
+
+	wTopic.RLock()
+	defer wTopic.RUnlock()
+
+	idStr, err := wTopic.Ring.Next(context.Background())
+	if err != nil {
+		return err
+	}
+
+	id, err := strconv.ParseInt(idStr, 10, 64)
+	if err != nil {
+		return err
+	}
+
+	wTopic.Bindings[id].Receiver() <- msg
+
+	return nil
 }
 
-// Unsubscribe a consumer from this topic.
-func (wTopic *WizardTopic) Unsubscribe(consumer string) {
+// Subscribe a Subscriber to this topic and receive a Subscription.
+func (wTopic *WizardTopic) Subscribe(sub Subscriber) (Subscription, error) {
 	wTopic.Lock()
+	defer wTopic.Unlock()
 
-	delete(wTopic.Bindings, consumer)
+	id := wTopic.counter
+	wTopic.Bindings[id] = sub
+	wTopic.counter++
+
+	if wTopic.Ring != nil {
+		if err := wTopic.Ring.Add(context.Background(), strconv.FormatInt(id, 10)); err != nil {
+			return Subscription{}, err
+		}
+	}
 
+	return Subscription{
+		id:     id,
+		cancel: wTopic.unsubscribe,
+	}, nil
+}
+
+// Unsubscribe a consumer from this topic.
+func (wTopic *WizardTopic) unsubscribe(id int64) {
+	wTopic.Lock()
+	delete(wTopic.Bindings, id)
+	if wTopic.Ring != nil {
+		// TODO: What do we do in the case of this failing?
```

Comment:
unsubscribe should return an error, and the error should be threaded through to Cancel.